# Prediction of temperature using machine learning and comparison of various regression models for Yarra Valley

The dataset that we have chosen for this project was obtained by the nasa portal https://power.larc.nasa.gov/data-access-viewer/¶<br>
Latitude:  -37.6327  Longitude:  145.7981<br>
Time Extent:   09/30/1981  -  12/31/2020<br>
Elevation:  393.26 meters<br>
Yarra Valley

##### We must first import the necessary libraries and modules for data preprocessing and model development.

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

We must first determine if the data is suitable before feeding it to the model. There may be some missing or null values in the data, as well as some non-required values that must be handled properly.
We should make an effort to comprehend the data thoroughly; this will aid in data processing.
As a result, our data consists of 15 columns and a large number of rows. 

In [2]:
yarra_df = pd.read_csv('data/YarraValley.csv')

In [3]:
yarra_df.head()

LAT        LON  YEAR  MO  DY  T2MDEW  T2MWET  T2M_MAX  T2M_MIN    T2M  \
0 -37.63269  145.79811  1981  10   1    6.80    6.80    14.85     7.13  11.42   
1 -37.63269  145.79811  1981  10   2   10.37   10.37    17.09    11.62  13.98   
2 -37.63269  145.79811  1981  10   3    5.53    5.53    12.16     4.72   7.49   
3 -37.63269  145.79811  1981  10   4    5.97    5.98    11.61     4.57   8.28   
4 -37.63269  145.79811  1981  10   5    7.24    7.24    16.00     8.31  11.68   

   PRECTOT  WS10M     PS  QV2M  ALLSKY_SFC_SW_DWN  
0     4.54   1.75  96.19  6.43             -999.0  
1    15.97   1.81  95.45  8.33             -999.0  
2    12.64   1.91  95.82  6.01             -999.0  
3     2.77   1.83  97.14  6.03             -999.0  
4     0.10   1.98  97.46  6.53             -999.0

The columns in the dataset are :

In [4]:
yarra_df.columns

Index(['LAT', 'LON', 'YEAR', 'MO', 'DY', 'T2MDEW', 'T2MWET', 'T2M_MAX',
       'T2M_MIN', 'T2M', 'PRECTOT', 'WS10M', 'PS', 'QV2M',
       'ALLSKY_SFC_SW_DWN'],
      dtype='object')

The shape of the dataset :

In [5]:
yarra_df.shape

(14337, 15)

Lets see the description of the datasset:

In [6]:
yarra_df.describe()

LAT           LON          YEAR            MO            DY  \
count  1.433700e+04  1.433700e+04  14337.000000  14337.000000  14337.000000   
mean  -3.763269e+01  1.457981e+02   2000.872358      6.551580     15.730557   
std    1.399818e-12  3.586945e-11     11.332154      3.456927      8.800849   
min   -3.763269e+01  1.457981e+02   1981.000000      1.000000      1.000000   
25%   -3.763269e+01  1.457981e+02   1991.000000      4.000000      8.000000   
50%   -3.763269e+01  1.457981e+02   2001.000000      7.000000     16.000000   
75%   -3.763269e+01  1.457981e+02   2011.000000     10.000000     23.000000   
max   -3.763269e+01  1.457981e+02   2020.000000     12.000000     31.000000   

             T2MDEW        T2MWET       T2M_MAX       T2M_MIN           T2M  \
count  14337.000000  14337.000000  14337.000000  14337.000000  14337.000000   
mean       7.287119      7.284904     17.842395      8.133656     12.666312   
std        3.205294      3.199686      7.032904      3.974373      5.321336   
min       -3.890000     -3.790000      3.300000     -2.670000      1.420000   
25%        5.040000      5.040000     12.130000      5.210000      8.330000   
50%        6.870000      6.870000     16.760000      7.610000     11.840000   
75%        9.110000      9.110000     22.830000     10.530000     16.280000   
max       21.830000     21.830000     42.350000     25.750000     32.650000   

            PRECTOT         WS10M            PS          QV2M  \
count  14337.000000  14337.000000  14337.000000  14337.000000   
mean       2.540024      1.447113     97.075005      6.763635   
std        4.801685      0.479854      0.677898      1.605244   
min        0.000000      0.430000     94.010000      2.910000   
25%        0.040000      1.090000     96.630000      5.650000   
50%        0.450000      1.360000     97.100000      6.420000   
75%        2.850000      1.720000     97.540000      7.490000   
max       70.660000      3.910000     99.160000     17.120000   

       ALLSKY_SFC_SW_DWN  
count       14337.000000  
mean          -40.793936  
std           207.653537  
min          -999.000000  
25%             2.120000  
50%             3.540000  
75%             5.840000  
max             9.690000

The mean Maximo temperature is approximately 18 degrees, with a huge variance of 7 degrees. We have entries for all of the columns for a given day, and the data is reliable.

# Data Mining

 ##### prints a concise sumary of a DataFrame

In [7]:
yarra_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14337 entries, 0 to 14336
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   LAT                14337 non-null  float64
 1   LON                14337 non-null  float64
 2   YEAR               14337 non-null  int64  
 3   MO                 14337 non-null  int64  
 4   DY                 14337 non-null  int64  
 5   T2MDEW             14337 non-null  float64
 6   T2MWET             14337 non-null  float64
 7   T2M_MAX            14337 non-null  float64
 8   T2M_MIN            14337 non-null  float64
 9   T2M                14337 non-null  float64
 10  PRECTOT            14337 non-null  float64
 11  WS10M              14337 non-null  float64
 12  PS                 14337 non-null  float64
 13  QV2M               14337 non-null  float64
 14  ALLSKY_SFC_SW_DWN  14337 non-null  float64
dtypes: float64(12), int64(3)
memory usage: 1.6 MB


Let's see if the dataset has any missing values.

In [8]:
yarra_df.isnull().any() # checks if there is any null value in a particular column

LAT                  False
LON                  False
YEAR                 False
MO                   False
DY                   False
T2MDEW               False
T2MWET               False
T2M_MAX              False
T2M_MIN              False
T2M                  False
PRECTOT              False
WS10M                False
PS                   False
QV2M                 False
ALLSKY_SFC_SW_DWN    False
dtype: bool

There are no values in any of the columns, as can be seen above.

Renaming few columns

In [9]:
yarra_df = yarra_df.rename(columns={'LAT':'latitude','LON':'longitude','YEAR':'year', 'MO':'month', 'DY':'day','PRECTOT':'precipitation', 'T2M_MAX':'max_temp', 'T2M_MIN':'min_temp', 'T2MWET':'Wet Bulb', 'T2MDEW':'Dew/Frost', 'QV2M':'Specific_Humidity' })
print(yarra_df.shape)
yarra_df

(14337, 15)


latitude  longitude  year  month  day  Dew/Frost  Wet Bulb  max_temp  \
0     -37.63269  145.79811  1981     10    1       6.80      6.80     14.85   
1     -37.63269  145.79811  1981     10    2      10.37     10.37     17.09   
2     -37.63269  145.79811  1981     10    3       5.53      5.53     12.16   
3     -37.63269  145.79811  1981     10    4       5.97      5.98     11.61   
4     -37.63269  145.79811  1981     10    5       7.24      7.24     16.00   
...         ...        ...   ...    ...  ...        ...       ...       ...   
14332 -37.63269  145.79811  2020     12   27       8.52      8.29     31.43   
14333 -37.63269  145.79811  2020     12   28       2.41      2.35     19.15   
14334 -37.63269  145.79811  2020     12   29       4.57      4.40     22.22   
14335 -37.63269  145.79811  2020     12   30       9.01      8.91     24.82   
14336 -37.63269  145.79811  2020     12   31      12.58     12.47     26.42   

       min_temp    T2M  precipitation  WS10M     PS  Specific_Humidity  \
0          7.13  11.42           4.54   1.75  96.19               6.43   
1         11.62  13.98          15.97   1.81  95.45               8.33   
2          4.72   7.49          12.64   1.91  95.82               6.01   
3          4.57   8.28           2.77   1.83  97.14               6.03   
4          8.31  11.68           0.10   1.98  97.46               6.53   
...         ...    ...            ...    ...    ...                ...   
14332      9.98  21.17           2.82   2.41  96.12               7.24   
14333      7.34  12.41           0.04   1.81  96.62               4.70   
14334      6.50  14.49           0.03   1.55  97.01               5.45   
14335      8.33  16.81           0.00   1.22  97.19               7.36   
14336     11.84  19.41           0.96   1.13  97.23               9.33   

       ALLSKY_SFC_SW_DWN  
0                -999.00  
1                -999.00  
2                -999.00  
3                -999.00  
4                -999.00  
...                  ...  
14332               2.67  
14333               8.71  
14334               9.22  
14335               8.19  
14336               5.87  

[14337 rows x 15 columns]

In [10]:
yarra_df.describe()

latitude     longitude          year         month           day  \
count  1.433700e+04  1.433700e+04  14337.000000  14337.000000  14337.000000   
mean  -3.763269e+01  1.457981e+02   2000.872358      6.551580     15.730557   
std    1.399818e-12  3.586945e-11     11.332154      3.456927      8.800849   
min   -3.763269e+01  1.457981e+02   1981.000000      1.000000      1.000000   
25%   -3.763269e+01  1.457981e+02   1991.000000      4.000000      8.000000   
50%   -3.763269e+01  1.457981e+02   2001.000000      7.000000     16.000000   
75%   -3.763269e+01  1.457981e+02   2011.000000     10.000000     23.000000   
max   -3.763269e+01  1.457981e+02   2020.000000     12.000000     31.000000   

          Dew/Frost      Wet Bulb      max_temp      min_temp           T2M  \
count  14337.000000  14337.000000  14337.000000  14337.000000  14337.000000   
mean       7.287119      7.284904     17.842395      8.133656     12.666312   
std        3.205294      3.199686      7.032904      3.974373      5.321336   
min       -3.890000     -3.790000      3.300000     -2.670000      1.420000   
25%        5.040000      5.040000     12.130000      5.210000      8.330000   
50%        6.870000      6.870000     16.760000      7.610000     11.840000   
75%        9.110000      9.110000     22.830000     10.530000     16.280000   
max       21.830000     21.830000     42.350000     25.750000     32.650000   

       precipitation         WS10M            PS  Specific_Humidity  \
count   14337.000000  14337.000000  14337.000000       14337.000000   
mean        2.540024      1.447113     97.075005           6.763635   
std         4.801685      0.479854      0.677898           1.605244   
min         0.000000      0.430000     94.010000           2.910000   
25%         0.040000      1.090000     96.630000           5.650000   
50%         0.450000      1.360000     97.100000           6.420000   
75%         2.850000      1.720000     97.540000           7.490000   
max        70.660000      3.910000     99.160000          17.120000   

       ALLSKY_SFC_SW_DWN  
count       14337.000000  
mean          -40.793936  
std           207.653537  
min          -999.000000  
25%             2.120000  
50%             3.540000  
75%             5.840000  
max             9.690000

In [11]:
yarra_df['date'] = pd.to_datetime(yarra_df[['year', 'month', 'day']])
cols = list(yarra_df.columns)
cols = [cols[-1]] + cols[:-1]

yarra_df = yarra_df[cols]

yarra_df

date  latitude  longitude  year  month  day  Dew/Frost  Wet Bulb  \
0     1981-10-01 -37.63269  145.79811  1981     10    1       6.80      6.80   
1     1981-10-02 -37.63269  145.79811  1981     10    2      10.37     10.37   
2     1981-10-03 -37.63269  145.79811  1981     10    3       5.53      5.53   
3     1981-10-04 -37.63269  145.79811  1981     10    4       5.97      5.98   
4     1981-10-05 -37.63269  145.79811  1981     10    5       7.24      7.24   
...          ...       ...        ...   ...    ...  ...        ...       ...   
14332 2020-12-27 -37.63269  145.79811  2020     12   27       8.52      8.29   
14333 2020-12-28 -37.63269  145.79811  2020     12   28       2.41      2.35   
14334 2020-12-29 -37.63269  145.79811  2020     12   29       4.57      4.40   
14335 2020-12-30 -37.63269  145.79811  2020     12   30       9.01      8.91   
14336 2020-12-31 -37.63269  145.79811  2020     12   31      12.58     12.47   

       max_temp  min_temp    T2M  precipitation  WS10M     PS  \
0         14.85      7.13  11.42           4.54   1.75  96.19   
1         17.09     11.62  13.98          15.97   1.81  95.45   
2         12.16      4.72   7.49          12.64   1.91  95.82   
3         11.61      4.57   8.28           2.77   1.83  97.14   
4         16.00      8.31  11.68           0.10   1.98  97.46   
...         ...       ...    ...            ...    ...    ...   
14332     31.43      9.98  21.17           2.82   2.41  96.12   
14333     19.15      7.34  12.41           0.04   1.81  96.62   
14334     22.22      6.50  14.49           0.03   1.55  97.01   
14335     24.82      8.33  16.81           0.00   1.22  97.19   
14336     26.42     11.84  19.41           0.96   1.13  97.23   

       Specific_Humidity  ALLSKY_SFC_SW_DWN  
0                   6.43            -999.00  
1                   8.33            -999.00  
2                   6.01            -999.00  
3                   6.03            -999.00  
4                   6.53            -999.00  
...                  ...                ...  
14332               7.24               2.67  
14333               4.70               8.71  
14334               5.45               9.22  
14335               7.36               8.19  
14336               9.33               5.87  

[14337 rows x 16 columns]

Setting the date column as index

In [12]:
yarra_df.set_index('date', inplace=True)



#### Getting the average tempereture from minimum and maximum temperature

In [13]:
col = yarra_df.loc[:, ('max_temp','min_temp')]
yarra_df['avg_temp'] = col.mean(axis=1)
yarra_df

latitude  longitude  year  month  day  Dew/Frost  Wet Bulb  \
date                                                                     
1981-10-01 -37.63269  145.79811  1981     10    1       6.80      6.80   
1981-10-02 -37.63269  145.79811  1981     10    2      10.37     10.37   
1981-10-03 -37.63269  145.79811  1981     10    3       5.53      5.53   
1981-10-04 -37.63269  145.79811  1981     10    4       5.97      5.98   
1981-10-05 -37.63269  145.79811  1981     10    5       7.24      7.24   
...              ...        ...   ...    ...  ...        ...       ...   
2020-12-27 -37.63269  145.79811  2020     12   27       8.52      8.29   
2020-12-28 -37.63269  145.79811  2020     12   28       2.41      2.35   
2020-12-29 -37.63269  145.79811  2020     12   29       4.57      4.40   
2020-12-30 -37.63269  145.79811  2020     12   30       9.01      8.91   
2020-12-31 -37.63269  145.79811  2020     12   31      12.58     12.47   

            max_temp  min_temp    T2M  precipitation  WS10M     PS  \
date                                                                 
1981-10-01     14.85      7.13  11.42           4.54   1.75  96.19   
1981-10-02     17.09     11.62  13.98          15.97   1.81  95.45   
1981-10-03     12.16      4.72   7.49          12.64   1.91  95.82   
1981-10-04     11.61      4.57   8.28           2.77   1.83  97.14   
1981-10-05     16.00      8.31  11.68           0.10   1.98  97.46   
...              ...       ...    ...            ...    ...    ...   
2020-12-27     31.43      9.98  21.17           2.82   2.41  96.12   
2020-12-28     19.15      7.34  12.41           0.04   1.81  96.62   
2020-12-29     22.22      6.50  14.49           0.03   1.55  97.01   
2020-12-30     24.82      8.33  16.81           0.00   1.22  97.19   
2020-12-31     26.42     11.84  19.41           0.96   1.13  97.23   

            Specific_Humidity  ALLSKY_SFC_SW_DWN  avg_temp  
date                                                        
1981-10-01               6.43            -999.00    10.990  
1981-10-02               8.33            -999.00    14.355  
1981-10-03               6.01            -999.00     8.440  
1981-10-04               6.03            -999.00     8.090  
1981-10-05               6.53            -999.00    12.155  
...                       ...                ...       ...  
2020-12-27               7.24               2.67    20.705  
2020-12-28               4.70               8.71    13.245  
2020-12-29               5.45               9.22    14.360  
2020-12-30               7.36               8.19    16.575  
2020-12-31               9.33               5.87    19.130  

[14337 rows x 16 columns]

##### Dropping the columns latitude and longitude  and adding the column region with Yarra Valley region.

In [14]:
yarra_df['region'] = yarra_df.latitude + yarra_df['longitude']
yarra_df.drop(['latitude', 'longitude'], axis=1, inplace=True)
yarra_df = yarra_df.replace(yarra_df['region'].values, 'YarraValley', regex=True,)
yarra_df

year  month  day  Dew/Frost  Wet Bulb  max_temp  min_temp    T2M  \
date                                                                           
1981-10-01  1981     10    1       6.80      6.80     14.85      7.13  11.42   
1981-10-02  1981     10    2      10.37     10.37     17.09     11.62  13.98   
1981-10-03  1981     10    3       5.53      5.53     12.16      4.72   7.49   
1981-10-04  1981     10    4       5.97      5.98     11.61      4.57   8.28   
1981-10-05  1981     10    5       7.24      7.24     16.00      8.31  11.68   
...          ...    ...  ...        ...       ...       ...       ...    ...   
2020-12-27  2020     12   27       8.52      8.29     31.43      9.98  21.17   
2020-12-28  2020     12   28       2.41      2.35     19.15      7.34  12.41   
2020-12-29  2020     12   29       4.57      4.40     22.22      6.50  14.49   
2020-12-30  2020     12   30       9.01      8.91     24.82      8.33  16.81   
2020-12-31  2020     12   31      12.58     12.47     26.42     11.84  19.41   

            precipitation  WS10M     PS  Specific_Humidity  ALLSKY_SFC_SW_DWN  \
date                                                                            
1981-10-01           4.54   1.75  96.19               6.43            -999.00   
1981-10-02          15.97   1.81  95.45               8.33            -999.00   
1981-10-03          12.64   1.91  95.82               6.01            -999.00   
1981-10-04           2.77   1.83  97.14               6.03            -999.00   
1981-10-05           0.10   1.98  97.46               6.53            -999.00   
...                   ...    ...    ...                ...                ...   
2020-12-27           2.82   2.41  96.12               7.24               2.67   
2020-12-28           0.04   1.81  96.62               4.70               8.71   
2020-12-29           0.03   1.55  97.01               5.45               9.22   
2020-12-30           0.00   1.22  97.19               7.36               8.19   
2020-12-31           0.96   1.13  97.23               9.33               5.87   

            avg_temp       region  
date                               
1981-10-01    10.990  YarraValley  
1981-10-02    14.355  YarraValley  
1981-10-03     8.440  YarraValley  
1981-10-04     8.090  YarraValley  
1981-10-05    12.155  YarraValley  
...              ...          ...  
2020-12-27    20.705  YarraValley  
2020-12-28    13.245  YarraValley  
2020-12-29    14.360  YarraValley  
2020-12-30    16.575  YarraValley  
2020-12-31    19.130  YarraValley  

[14337 rows x 15 columns]

### Dropping unnecesary columns
We will drop year, month and day columns as we have extracted that info to create the "date" column. Also, we will drop ALLSKY_SFC_SW_DWN, precipitation, latitud and longitude as they are always the same values because we are talking about ame place, Yarra Valley.

In [15]:
yarra_df.drop(['year','month','day','ALLSKY_SFC_SW_DWN', 'precipitation','min_temp','max_temp'], axis='columns', inplace=True)
yarra_df

Dew/Frost  Wet Bulb    T2M  WS10M     PS  Specific_Humidity  \
date                                                                      
1981-10-01       6.80      6.80  11.42   1.75  96.19               6.43   
1981-10-02      10.37     10.37  13.98   1.81  95.45               8.33   
1981-10-03       5.53      5.53   7.49   1.91  95.82               6.01   
1981-10-04       5.97      5.98   8.28   1.83  97.14               6.03   
1981-10-05       7.24      7.24  11.68   1.98  97.46               6.53   
...               ...       ...    ...    ...    ...                ...   
2020-12-27       8.52      8.29  21.17   2.41  96.12               7.24   
2020-12-28       2.41      2.35  12.41   1.81  96.62               4.70   
2020-12-29       4.57      4.40  14.49   1.55  97.01               5.45   
2020-12-30       9.01      8.91  16.81   1.22  97.19               7.36   
2020-12-31      12.58     12.47  19.41   1.13  97.23               9.33   

            avg_temp       region  
date                               
1981-10-01    10.990  YarraValley  
1981-10-02    14.355  YarraValley  
1981-10-03     8.440  YarraValley  
1981-10-04     8.090  YarraValley  
1981-10-05    12.155  YarraValley  
...              ...          ...  
2020-12-27    20.705  YarraValley  
2020-12-28    13.245  YarraValley  
2020-12-29    14.360  YarraValley  
2020-12-30    16.575  YarraValley  
2020-12-31    19.130  YarraValley  

[14337 rows x 8 columns]

In [16]:
yarra_df['avg_temp'].value_counts()

9.285     14
8.240     13
9.230     13
8.790     12
5.365     12
          ..
7.850      1
7.265      1
28.620     1
17.310     1
17.525     1
Name: avg_temp, Length: 5361, dtype: int64

Let's now isolate the function that needs to be expected from the rest of the features. The remainder of the dataset is stored in yarra x, while the temperature column is stored in weather y.

In [18]:
yarra_df_num = yarra_df[list(yarra_df.dtypes[yarra_df.dtypes!='object'].index)]

Now that we've prepared our dataset, it's time to feed it to the model for training.

In [19]:
yarra_y = yarra_df_num.pop('avg_temp')
yarra_x = yarra_df_num

In [20]:
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LinearRegression

import tensorflow as tf
from tensorflow import keras

from sklearn import preprocessing

Now is the time to divide the dataset into training and testing parts.

In [21]:
train_x,test_x,train_y,test_y = train_test_split(yarra_x, yarra_y,test_size = 0.2, random_state=4)

In [22]:
train_x.head()

Dew/Frost  Wet Bulb    T2M  WS10M     PS  Specific_Humidity
date                                                                   
2007-09-09       6.45      6.46  10.23   1.51  97.13               6.23
2004-04-05      10.45     10.45  17.88   1.48  97.86               8.09
2002-11-24       8.43      8.46  22.79   1.66  96.50               7.69
2001-12-02       8.16      8.17  16.37   1.29  96.29               7.17
1997-03-18       4.75      4.77  18.45   1.52  96.96               5.57

Train x has all of the features except the temperature average, and train y has the temperature range for those features. In supervised machine learning, we feed the model input and associated output first, then test it with new data.

## Implementation of the Models 

### Linear Regression

In [23]:
model = LinearRegression()
model.fit(train_x,train_y)

LinearRegression()

In [24]:
prediction = model.predict(test_x)

In [25]:
from sklearn import metrics
from math import sqrt

def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

mape = mean_absolute_percentage_error(test_y, prediction)
print(mape)

3.3101321425181887


For this model the was Error — 3.31013 

In [26]:
pd.DataFrame({'actual':test_y,
             'prediction':prediction,
             'diff':(test_y-prediction)})

actual  prediction      diff
date                                    
2017-05-01  11.825   11.929435 -0.104435
1998-11-01   9.080    9.297473 -0.217473
2000-06-16   9.380    9.897217 -0.517217
2008-09-16   6.655    6.454141  0.200859
1997-01-23  16.650   16.349373  0.300627
...            ...         ...       ...
2020-01-04  20.185   21.114300 -0.929300
1991-12-10  11.550   12.807768 -1.257768
1997-04-02  12.455   12.031206  0.423794
2006-06-15   7.520    7.825107 -0.305107
1996-02-10  14.035   13.061734  0.973266

[2868 rows x 3 columns]

As we can see the accuracy of the temperature is very similar to what was predicted and as an exemple of the first row  the temperature in 01/05/17 was **11.825** and the prediction was **11.92** with a difference of only 0.10 less than it was.

# Random Forest with maximum Dept — 10

In [61]:
from sklearn.ensemble import RandomForestRegressor
regr = RandomForestRegressor(max_depth=10,random_state=0, n_estimators=100)
regr.fit(train_x,train_y)

RandomForestRegressor(max_depth=10, random_state=0)

In [51]:
prediction2 = regressor.predict(test_x)

In [57]:
def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

mape = mean_absolute_percentage_error(test_y, prediction2)
print(mape)

4.470773544888394


For this model the was Error — 4.47077 

In [53]:
pd.DataFrame({'actual':test_y,
             'prediction':prediction2,
             'diff':(test_y-prediction2)})

actual  prediction   diff
date                                 
2017-05-01  11.825      12.015 -0.190
1998-11-01   9.080       8.605  0.475
2000-06-16   9.380       9.570 -0.190
2008-09-16   6.655       6.780 -0.125
1997-01-23  16.650      16.275  0.375
...            ...         ...    ...
2020-01-04  20.185      20.990 -0.805
1991-12-10  11.550      12.935 -1.385
1997-04-02  12.455      12.175  0.280
2006-06-15   7.520       7.930 -0.410
1996-02-10  14.035      13.145  0.890

[2868 rows x 3 columns]

As we can see the accuracy of the temperature is very similar to what was predicted and as an exemple of the first row  the temperature on 01/05/17 was **11.825** and the prediction was **12.015** with a difference of only 0.19 less than it was.

# Comparison:

Linear Regression = 3.31013 <br>

Random Forest Dept — 10 => 4.470


We can see that Linear Regression is the best model to predict the weather on this datset.